src/main/java/org/wildfly/security/auth/spi/Attributes.java

```diff
@@ -576,13 +598,81 @@ default boolean isEmpty() {
     }
 
     /**
-     * Get the collection of values for the given key.  The result may implement {@link SetEntry} if the values
-     * are distinct (for example, a role or group set).
+     * Returns a read-only instance of this instance.
      *
-     * @param key the attribute name
-     * @return the (possibly empty) attribute value collection
+     * @return a read-only instance of this instance.
      */
-    Entry get(String key);
+    default Attributes asReadOnly() {
+        return new Attributes() {
+            private Map<String, Entry> entryCache;
+            private Collection<Entry> cachedEntries;
+
+            @Override
+            public Collection<Entry> entries() {
+                if (this.cachedEntries != null) {
+                    return this.cachedEntries;
+                }
+
+                return this.cachedEntries = new AbstractCollection<Entry>() {
```

Comment:
`Attributes.super.values();` returns a `Collections<String>`, right ? And we are expecting to return a  `Collection<Entry>`.